cstrml/market/src/lib.rs

```diff
@@ -429,16 +461,71 @@ impl<T: Trait> Module<T> {
                     T::Payment::pay_sorder(order_id);
                 }
 
-                // TODO: add market slashing here
-                // TODO: Record `failed_count` from `Success` to `Failed`
-                // TODO: Record `failed_duration` from `Failed` to `Success`
-
                 // 2. Update storage order
                 <StorageOrders<T>>::insert(order_id, so);
             }
         }
     }
 
+    pub fn update_slash_record(order_id: &<T as system::Trait>::Hash) {
+        // 1. Update record's count
+        let so = Self::storage_orders(order_id).unwrap();
+        <OrderSlashRecords<T>>::mutate(order_id, |status_count| {
+            if let Some(sc) = status_count {
+                if so.status == OrderStatus::Success {
+                    sc.success += 1;
+                } else if so.status == OrderStatus::Failed {
+                    sc.failed += 1;
+                } else {}
+            }
+
+        });
+
+        // 2. Maybe do slash
+        Self::maybe_slash_provider(&order_id, &so);
+    }
+
+    pub fn maybe_slash_provider(order_id: &T::Hash, so: &StorageOrder<T::AccountId, BalanceOf<T>>) {
+        let mut status_count = Self::order_slash_records(order_id).unwrap();
+        let slash_duration = T::SlashingDuration::get();
+        if status_count.success + status_count.failed >= slash_duration && status_count.can_slash {
+            // Calculate slash value
+            let mut slash_value = Zero::zero();
+            if status_count.success >= ((slash_duration as f64) * 0.95) as EraIndex {
+            } 
+            else if status_count.success >= ((slash_duration as f64) * 0.90) as EraIndex {
+                slash_value =
+                <T::CurrencyToBalance as Convert<u128, BalanceOf<T>>>::
+                convert(<T::CurrencyToBalance as Convert<BalanceOf<T>, u128>>::
+                convert(so.amount) / 2 as u128);
+ 
+            } 
+            else {
+                slash_value = so.amount;
+            }
+
+            // Do slash
+            if !slash_value.is_zero() {
+                slash_value = slash_value.min(so.amount - status_count.slashed_value);
+                status_count.slashed_value += slash_value;
+                if status_count.slashed_value >= so.amount {
+                    status_count.can_slash = false;
```

Comment:
What if the `can_punish` be false? Will this influent delayed payment? like If the `remain_value`(pledge - punished_value) < `sorder_amount`, payment should failed

Reply:
it won't have impact on delayed payment.